Support monitoring via opencensus

Makpoc opened the pull request:
Fixes #11

Makpoc committed:
Support opensensus for monitoring

Makpoc committed:
Add opensensus + move go-kit to separate file

Makpoc committed:
Create operation tag and add it to context

Makpoc committed:
Pull WriteSource out of models

Makpoc committed:
Change measurments to pointer to measurements

Makpoc committed:
Add examples with generated files

Makpoc committed:
Add a context decorator func

Change the generated monitoring structure to receive a context
decorator which will allow clients to inject additional tags.

Makpoc committed:
Update mongen examples and simplify them for both providers

Makpoc committed:
Add error handling for generated code

The criteria for MustNewKey to panic is if the key doesn't meet any of
the following properties:

* len > 0
* len < 255
* !ASCII

For `MustNewKey` we control the key so it will not panic

For `New` however it _may_ happen if the name of the method is too long
(over 255) or it contains non-ascii characters.

Makpoc committed:
Split gokit and OC implementations in packages

Additionally:
* Un-export structs and functions that do not need to be public
* Add documentation for public functions and methods

Bo0mer commented:
Hello and thanks for the PR!

I'll try it out and will post any comments that I have regarding the change.

Thanks,
Ivan

Bo0mer commented on pkg/astgen/method.go:
```diff
@@ -26,6 +26,11 @@ func (m *Method) SetType(funcType *ast.FuncType) {
 	m.statements = append(m.statements, stmt)
 }
+func (m *Method) AddStatements(stmts []ast.Stmt) {
+	for _, stmt := range stmts {
```
You could rather do:
```
	m.statements = append(m.statements, stmts...)
```

Bo0mer commented on cmd/mongen/internal/opencensus/oc_model.go:
```diff
@@ -0,0 +1,77 @@
+	fileBuilder *astgen.File
+	structName  string
+
+	packageAliases packageAliases
```
WDYT about embedding this, so it's shorter to use the members? I.e.:

```go
// packageAliases holds the aliases of all imported packages in the generated source file.
type packageAliases struct {
	timePkg    string
	contextPkg string
	tagPkg     string
	statsPkg   string
	sourcePkg  string
}

type opencensusModel struct {
	fileBuilder *astgen.File
	structName  string

	packageAliases
}

func newOpencensusModel(interfacePath, interfaceName, structName, targetPkg string) *opencensusModel {
	file := astgen.NewFile(targetPkg)
	m := &opencensusModel{
		fileBuilder: file,
		structName:  structName,
		packageAliases: packageAliases{
			timePkg:    file.AddImport("", "time"),
			contextPkg: file.AddImport("", "context"),
			tagPkg:     file.AddImport("", "go.opencensus.io/tag"),
			statsPkg:   file.AddImport("", "go.opencesus.io/stats"),
			sourcePkg:  file.AddImport("", interfacePath),
		},
	}

	strct := astgen.NewStruct(structName)
	strct.AddField("next", m.sourcePkg, interfaceName)
	strct.AddFieldWithType(totalOps, pointerExpr(m.statsPkg, "Int64Measure"))
	strct.AddFieldWithType(failedOps, pointerExpr(m.statsPkg, "Int64Measure"))
	strct.AddFieldWithType(opsDuration, pointerExpr(m.statsPkg, "Float64Measure"))
	strct.AddFieldWithType(ctxFuncName, buildCtxFuncType(m.contextPkg))
	file.AppendDeclaration(strct)

	constructorBuilder := newOCConstructorBuilder(
		m.statsPkg, m.contextPkg, m.sourcePkg, interfaceName)
	file.AppendDeclaration(constructorBuilder)

	return m
}
```

Other changes that I've introduced are:
* packages are sorted so std lib ones are first
* the model struct is initialised in a single construct

Makpoc replied:
Embedding does make it shorter, but I like the explicit grouping under packageAliases when accessing the *Pkg properties.

:+1: for sorting the imports and changes to init.

Let me know how you feel about leaving the packageAliases as separate struct member tho.

Bo0mer replied:
Sure, I'm fine. Explicit is better than implicit.

Makpoc replied:
Cool. I've applied your other comments
* sort imports
* init the model in a single construct

Bo0mer commented on cmd/mongen/internal/common/builders.go:
```diff
@@ -0,0 +1,105 @@
+package common
```
I'm not a huge fan of `common` packages, as they tend to become bloated with everything. Do you find it reasonable to keep this code in the `cmd/mongen`'s `main` package?

Makpoc replied:
I understand. The `common` here was intended to mark the builders that are common for both `gokit` and `opencensus` and not as a _common dumping ground_. I'll see how i can move things around to remove the need for this package.

Makpoc replied:
`common` is not split into `commonbuilders` and `pkg/transformation`

Makpoc committed:
Apply code review comments

* Split `common` package into
  * `mongen/commonbuilders` containing the builders that are
used for both gokit and opencensus implementations
  * `pkg/transformation`, used by all generator tools.
* Sort imports
* Remove some unused functions and variables
* Fix review comment
https://github.com/Bo0mer/gentools/pull/14#discussion_r350584330
